ENH: integrate.trapezoid: add array API standard support

Armavica opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-20930.-->

Towards [gh-20930](https://github.com/scipy/scipy/issues/20930).

#### What does this implement/fix?
<!--Please explain your changes.-->

Implement the Array API for `integrate.trapezoid`.

#### Additional information
<!--Any additional information you think is important.-->

Hello! I met @lucascolley at the NumFocus summit this week, and he advised me to start with `integrate.trapezoid` to help with the Array API implementation effort. I am submitting this as a draft PR because I don't understand yet everything that is going on here, and I am looking for feedback and advice.

I have several questions so far:
* There is a masked array test, which does not work with Array API. I understand that supporting masked arrays is out of scope of Array API. How should we then proceed with this test?
* I remember that Lucas told me that `np.diff` would probably need to be reimplemented, so I don't really understand why my initial strategy of blindly replacing `np` with `xp` seems to pass some tests.

Armavica committed:
ENH: integrate.trapezoid: add array API standard support

lucascolley commented on scipy/integrate/_quadrature.py:
```diff
@@ -141,9 +144,9 @@ def trapezoid(y, x=None, dx=1.0, axis=-1):
-        ret = np.add.reduce(d * (y[tuple(slice1)]+y[tuple(slice2)])/2.0, axis)
+        d = xp.asarray(d)
+        y = xp.asarray(y)
+        ret = xp.add.reduce(d * (y[tuple(slice1)]+y[tuple(slice2)])/2.0, axis)
```
`add.reduce` isn't in the standard, I think that syntax is specific to NumPy ufuncs. I think we can replace it here with `xp.sum` though (which will also add along `axis`: https://data-apis.org/array-api/latest/API_specification/generated/array_api.sum.html).

lucascolley reviewed (commented):
thanks for working on this @Armavica !

> There is a masked array test, which does not work with Array API. I understand that supporting masked arrays is out of scope of Array API. How should we then proceed with this test?

You can decorate the test with `skip_xp_invalid_arg` from `scipy.conftest` - see https://scipy.github.io/devdocs/dev/api-dev/array_api.html#adding-tests.

> I remember that Lucas told me that np.diff would probably need to be reimplemented, so I don't really understand why my initial strategy of blindly replacing np with xp seems to pass some tests.

Let’s see if we get any failures when you push the converted tests :)

Armavica commented:
Thank you for your comments and for the pointers to documentation! I addressed your comments, it looks like tests are passing :)

lucascolley commented on scipy/integrate/_quadrature.py:
```diff
@@ -119,19 +121,20 @@ def trapezoid(y, x=None, dx=1.0, axis=-1):
         d = dx
     else:
-        x = np.asanyarray(x)
+        x = xp.asanyarray(x)
```
this can be `_asarray(..., subok=True)` too

lucascolley reviewed (changes requested):
We also need to convert the tests here as described on https://scipy.github.io/devdocs/dev/api-dev/array_api.html#adding-tests. At the moment, all of the tests still only input NumPy arrays, which is why `xp.diff` and `xp.asanyarray` aren't failing.

Armavica commented:
Ah, that's why! I had the feeling something was still wrong despite tests passing. Thank you, I get it now.

Armavica committed:
BUG: integrate.trapezoid: fix cases where 1 < x.ndim < y.ndim

Armavica commented:
I fixed the array API commit, and I added a commit to fix a behavior which was already present in the original version, in cases where `1 < x.ndim < axis < y.ndim`.

lucascolley commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -265,32 +266,34 @@ def test_zero_len_y(self):
 class TestTrapezoid:
-    def test_simple(self):
-        x = np.arange(-10, 10, .1)
-        r = trapezoid(np.exp(-.5 * x ** 2) / np.sqrt(2 * np.pi), dx=0.1)
+    @array_api_compatible
```
you can apply `@array_api_compatible` just once to the `TestTrapezoid` class rather than to each method since all tests have been converted. (you'll need to add an unused `xp` arg to `test_masked` I think, but that's fine)

lucascolley commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -265,32 +266,34 @@ def test_zero_len_y(self):
+        x = xp.arange(-10, 10, .1)
+        r = trapezoid(xp.exp(-.5 * x ** 2) / xp.sqrt(2 * xp.asarray(xp.pi)), dx=0.1)
         # check integral of normal equals 1
         assert_allclose(r, 1)
```
we want to use `xp_assert_close` from `scipy._lib._array_api` instead of `assert_allclose` from `np.testing` here. It comes with additional strictness checks, the most relevant one here being a check that the namespaces match. You may have to grapple with whether we are expecting arrays or scalars, too, after switching to the new assertions.

(this could probably be made clearer in our array API dev docs, though!)

lucascolley reviewed (changes requested):
thanks, getting close!

Armavica commented:
I added the line to the file as requested.

It looks like this change automatically triggered review requests?

lucascolley commented:
yep, no worries, that is https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners

lucascolley committed:
TST: integrate.trapezoid: minor fixes

[skip cirrus] [skip circle]

lucascolley reviewed (approved):
thanks @Armavica ! I pushed a small update (explained inline) and I'm happy with this now. Will leave open for a while in case anyone has comments about the bug-fix!

Armavica commented:
Thank you for your help on this @lucascolley!

Armavica commented:
Note that another approach for what I fixed would be to do like in `simpson`:

https://github.com/scipy/scipy/blob/8e4d71a648c8c755310c518b2b0d63569bdd8918/scipy/integrate/_quadrature.py#L459-L461

lucascolley reviewed (changes requested):
Actually, one more thing! Could you add an array-like test, along the lines of the tests added in gh-21209? It doesn't have to be complicated, just a sanity check that we are still accepting Python lists.

Armavica committed:
TST: integrate.trapezoid: add array-like test

lucascolley commented:
I think it is just over-indented! It's currently sitting inside `test_masked`.

Armavica commented:
Gosh that's embarrassing… :sweat_smile: It's fixed and the test seems to run now.

lucascolley reviewed (approved):
LGTM, thanks again!

lucascolley commented:
> Note that another approach for what I fixed would be to do like in simpson

What was the behaviour before your fix, by the way? Just nonsense returned?

mdhaber commented:
@Armavica thanks for working on this.

@lucascolley I saw that you requested my review. I took a close look and didn't find much wrong with the Array API translation, but mostly because I was distracted by the function's nonstandard broadcasting rules. They don't follow either those of NumPy linalg or `scipy.stats` (which differ only for positive `axis` when the arrays don't have the same number of dimensions), and they aren't documented. I'd suggest that as a follow-up we emit a `FutureWarning` for input arrays that are not of the same shape. Afterwards, this function could be greatly simplified by following standard broadcasting rules.

The one thing I did notice is that this preserved `dtype` before but doesn't now. This is because it uses `sum`, and in the 2022.12 of the array API standard (which is still in effect in `array_api_compat`), `sum` doesn't maintain dtype. Until 2023.12 takes effect, I'd suggest taking the `result_type` and passing that explicitly to `sum`.

Armavica commented:
> What was the behaviour before your fix, by the way? Just nonsense returned?

No, an indexing error.

Armavica commented:
> The one thing I did notice is that this preserved `dtype` before but doesn't now. This is because it uses `sum`, and in the 2022.12 of the array API standard (which is still in effect in `array_api_compat`), `sum` doesn't maintain dtype. Until 2023.12 takes effect, I'd suggest taking the `result_type` and passing that explicitly to `sum`.

@mdhaber I added this. It turned out to break the previous version of `test_array_like`, because the input had an integer dtype, and it looks like `xp.sum` makes rounding errors in this case (returning 20 instead of the correct answer 22). So I also changed the test from `y = [t*t for t in x]` to `y = [float(t*t) for t in x]`.
Perhaps another option would be to let the sum return a float, that we could only then convert to the original dtype, resulting in possibly fewer rounding errors even on integer inputs?

mdhaber commented:
Thanks; I wasn't explicit enough. Please pass the arrays into `xp_broadcast_promote` to determine the result type. (Unfortunately, you cannot use the broadcasted arrays that are returned because `trapezoid` does not follow normal broadcasting rules. You can apply that `dtype` to the arrays with `xp.astype(..., copy=False)`.)

lucascolley committed:
flush ci

[skip cirrus]

lucascolley reviewed (approved):
CI failures look unrelated - thanks again @Armavica !

lucascolley commented on scipy/integrate/_quadrature.py:
```diff
@@ -119,31 +121,39 @@ def trapezoid(y, x=None, dx=1.0, axis=-1):
-        d = dx
-    else:
-        x = np.asanyarray(x)
-        if x.ndim == 1:
```
looks like this `x.ndim == 1` guard is what has gone missing for gh-21908.
